client/securedrop_client/gui/widgets.py

```diff
@@ -690,20 +857,94 @@ def show_sources(self, sources: list[Source]) -> None:
                 # Then call the function to remove the wrapper and its children.
                 self.delete_conversation(source_uuid)
 
+        # Show the correct conversation pane gui element depending on
+        # the number of sources a) available and b) selected.
+        # An improved approach will be to create an `on_source_context_update`
+        # pyQtSlot that subscribes/listens for storage updates and calls
+        # `show_sources` and `show_conversation_context`.
+        self._on_update_conversation_context()
+
+    def _on_update_conversation_context(self) -> None:
+        """
+        Show the correct view type based on the number of available and selected sources.
+
+        If there are no sources, show the empty conversation view.
+        If there are sources, but none are selected, show the "Select a source" view.
+        If there are sources and exactly one has been selected, show the conversation
+        with that source.
+        If multiple sources are selected, show the "Multiple sources selected" view.
+
+        This method can be triggered by a click event (list index changed) or by a "sync"
+        event (sourcelist updated), which redraws the list.
+
+        In the latter case, supply list[Source] and set is_redraw_event to True.
+
+        Return number of selected sources.
+        """
+        selected = len(self.source_list.selectedItems())
+        if selected == 0 and self.source_list.count() == 0:
+            self.view_layout.setCurrentIndex(self.NO_SOURCES_INDEX)
+        elif selected == 0:
+            self.view_layout.setCurrentIndex(self.NOTHING_SELECTED_INDEX)
+        elif selected > 1:
+            self.view_layout.setCurrentIndex(self.MULTI_SELECTED_INDEX)
+        else:
+            # Exactly one source selected
+            self.view_layout.setCurrentIndex(self.CONVERSATION_INDEX)
+
+    @pyqtSlot()
     def on_source_changed(self) -> None:
         """
-        Show conversation for the selected source.
+        Show conversation for the selected source, or, if multiple sources are selected,
+        show multi select view.
+        """
+
+        selected = len(self.source_list.selectedItems())
+        if selected == 1:
+            # One source selected; prepare the conversation widget
+            try:
+                source = self.source_list.get_selected_source()
+                if not source:
+                    return
```

Comment:
Is it possible to have no `source` with `selected == 1`?  Either way, isn't that a fault we should at least log, if not error on or try to recover from?

Reply:
I added a comment to explain this check - `get_selected_source` immediately above is racy (not a regression - historically racy), because it tries to refresh the source object, so yes it's possible, although very difficult to introduce, for the user to click a sourcelist object in the gui at the same time that the object is being deleted from the db.